agent/stats/engine.go

```diff
@@ -108,31 +108,51 @@ func (engine *DockerStatsEngine) MustInit(taskEngine ecsengine.TaskEngine, clust
 	return engine.Init()
 }
 
-// Init initializes the docker client's event engine. This must be called
-// to subscribe to the docker's event stream.
+// Init subscribes to the container change event stream.
 func (engine *DockerStatsEngine) Init() error {
-	engine.ctx, engine.unsubscribeContainerEvents = context.WithCancel(context.TODO())
-	err := engine.openEventStream()
+	// Subscribe to the container change event stream
+	err := engine.containerChangeEventStream.Subscribe(containerChangeHandler, engine.handleDockerEvents)
 	if err != nil {
-		return err
+		return fmt.Errorf("Failed to subscribe to container change event stream, err %v", err)
 	}
 
 	go engine.listContainersAndStartEventHandler()
+	go func() {
+		// Waiting for the event stream to close
+		ctx := engine.containerChangeEventStream.Context()
+		c, _ := context.WithCancel(ctx)
```

Comment:
I don't believe you need a child context if you're not closing it independently of the parent.  You should be able to just use `<-ctx.Done()` directly without making a new one.

> Done returns a channel that's closed when work done on behalf of this context should be canceled. Done may return nil if this context can never be canceled. Successive calls to Done return the same value.

(https://golang.org/pkg/context/#Context)